### PR TITLE
Fixes for building with LibreSSL

### LINKED ISSUE #1307: [openbsd] build fails while building SSLUtils.cc
Build fails on OpenBSD.

```
kern.version=OpenBSD 6.0-current (GENERIC.MP) #0: Wed Nov 23 20:54:10 MST 2016
    build@amd64.openbsd.org:/usr/src/sys/arch/amd64/compile/GENERIC.MP
```

```
c++ -DHAVE_CONFIG_H -I. -I../../lib  -I../../iocore/eventsystem -I../../iocore/net -I../../iocore/aio -I../../iocore/hostdb -I../../iocore/cache -I../../iocore/cluster -I../../iocore/utils -I../../iocore/dns -I../../lib -I../../lib/records -I../../proxy -I../../proxy/hdrs -I../../proxy/shared -I../../proxy/logging -I../../mgmt -I../../mgmt/utils -I../../proxy/api/ts -I../../proxy/http -I/usr/include -I/usr/local/include-Dopenbsd -D_LARGEFILE64_SOURCE=1 -D_COMPILE64BIT_SOURCE=1 -D_GNU_SOURCE -D_REENTRANT -D__STDC_LIMIT_MACROS=1 -D__STDC_FORMAT_MACROS=1 -D_FORTIFY_SOURCE=2 -I/usr/include -DOPENSSL_NO_SSL_INTERN -I/usr/local/include/tcl8.5  -O2 -pipe -O0 -g -std=c++11 -Wall -Wextra -Wno-ignored-qualifiers -Wno-unused-parameter -feliminate-unused-debug-symbols -fno-strict-aliasing -Wno-invalid-offsetof -fPIE -fstack-protector -mcx16 -MT SSLUtils.o -MD -MP -MF $depbase.Tpo -c -o SSLUtils.o SSLUtils.cc &&\
mv -f $depbase.Tpo $depbase.Po
SSLUtils.cc: In function 'void SSLInitializeLibrary()':
SSLUtils.cc:917:83: error: invalid conversion from 'void* (*)(size_t, const char*, int) {aka void* (*)(long unsigned int, const char*, int)}' to 'void* (*)(size_t) {aka void* (*)(long unsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 1 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:917:83: error: invalid conversion from 'void* (*)(void*, size_t, const char*, int) {aka void* (*)(void*, long unsigned int, const char*, int)}' to 'void* (*)(void*, size_t) {aka void* (*)(void*, longunsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 2 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:917:83: error: invalid conversion from 'void (*)(void*, const char*, int)' to 'void (*)(void*)' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 3 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void* (*)(size_t, const char*, int) {aka void* (*)(long unsigned int, const char*, int)}' to 'void* (*)(size_t) {aka void* (*)(long unsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 1 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void* (*)(void*, size_t, const char*, int) {aka void* (*)(void*, long unsigned int, const char*, int)}' to 'void* (*)(void*, size_t) {aka void* (*)(void*, longunsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 2 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void (*)(void*, const char*, int)' to 'void (*)(void*)' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 3 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:941:34: error: 'SSL_pthreads_thread_id' was not declared in this scope
     CRYPTO_THREADID_set_callback(SSL_pthreads_thread_id);
                                  ^
SSLUtils.cc: In function 'SSL_CTX* SSLInitServerContext(const SSLConfigParams*, const ssl_user_config*, Vec<x509_st*>&)':
SSLUtils.cc:1585:56: error: invalid conversion from 'SSL_SESSION* (*)(SSL*, const unsigned char*, int, int*) {aka ssl_session_st* (*)(ssl_st*, const unsigned char*, int, int*)}' to 'SSL_SESSION* (*)(ssl_st*, unsigned char*, int, int*) {aka ssl_session_st* (*)(ssl_st*, unsigned char*, int, int*)}' [-fpermissive]
     SSL_CTX_sess_set_get_cb(ctx, ssl_get_cached_session);
                                                        ^
In file included from P_SSLNetVConnection.h:41:0,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/ssl.h:932:6: note: initializing argument 2 of 'void SSL_CTX_sess_set_get_cb(SSL_CTX*, SSL_SESSION* (*)(ssl_st*, unsigned char*, int, int*))'
 void SSL_CTX_sess_set_get_cb(SSL_CTX *ctx,
      ^
```

### PR DESCRIPTION
**Pull request by jablko**
Fixes #1307

### CODE REVIEW

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1311/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1205/ for details.

**Comment by jirib** [comment]
It passes previously failing part on OpenBSD.

File: example/cppapi/websocket/WSBuffer.cc [hunk 1]
```diff
@@ -157,7 +157,7 @@ WSBuffer::read_buffered_message(std::string &message, int &code)
 std::string
 WSBuffer::ws_digest(std::string const &key)
 {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
```
Comment (bryancall) [inline example/cppapi/websocket/WSBuffer.cc]: Shouldn't you check against a version number for LIBRESSL_VERSION_NUMBER?

Comment (jpeach) [inline example/cppapi/websocket/WSBuffer.cc]: Do we really need all this ifdeffery? Can we just unconditionally use the modern API?

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1313/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1207/ for details.

File: example/cppapi/websocket/WSBuffer.cc (hunk 1 above)
Comment (jablko) [inline example/cppapi/websocket/WSBuffer.cc]: Good idea. Is it okay if I make that change in #1313?

Comment (jablko) [inline example/cppapi/websocket/WSBuffer.cc]: I don't know which future version of LibreSSL will change this behavior ...

Comment (bryancall) [inline example/cppapi/websocket/WSBuffer.cc]: True :)

**Comment by bryancall** [comment]
Is this pull request needed anymore after #1313 ?

**Comment by jirib** [comment]
Any news about merging LibreSSL support?

**Comment by jablko** [comment]
@bryancall Thanks for reviewing this! I'd like to land both (this and #1313).

I think this is ready to go? If you approve, I'll merge it.

#1313 depends on this PR, so if you approve that one, I would land both of them.

**Comment by bryancall** [comment]
@jablko It looks like there is going to be a merge conflict if you commit this and then #1313.  That is what I am seeing when I am testing it on my server:

patching file example/cppapi/websocket/WSBuffer.cc
Reversed (or previously applied) patch detected!  Assume -R? [n]

**Comment by jablko** [comment]
Thanks @bryancall! Merged.

Yeah, until this PR landed, it was also part of #1313 (is there a way around that?) so I guess applying both would conflict. Git is smart enough to match the commit in #1313 with this one, so the following works without a conflict. I don't know if there's a better way to do this ...

    $ git fetch https://github.com/jablko/trafficserver.git libressl
    $ git rebase --onto master master FETCH_HEAD
    $ git fetch https://github.com/jablko/trafficserver.git autoconf
    $ git rebase --onto master master FETCH_HEAD

**Pull request merged**